logstash-core/src/main/java/org/logstash/config/ir/graph/Graph.java

```diff
@@ -279,14 +279,21 @@ public void validate() throws InvalidIRException {
             throw new InvalidIRException("Graph has no leaf vertices!\n" + this.toString());
         }
 
-        List<List<Vertex>> duplicates = verticesByHash().values().stream().filter((group) -> group.size() > 1).collect(Collectors.toList());
-        if (!duplicates.isEmpty()) {
-            Stream<String> errorMessageGroups = duplicates.stream().
-                    map((group) -> group.stream().map(Object::toString).collect(Collectors.joining("===")));
-
-            String joinedErrorMessageGroups = errorMessageGroups.collect(Collectors.joining("\n---\n"));
-
-            throw new InvalidIRException("Some nodes on the graph are fully redundant!\n" + this + "|" + joinedErrorMessageGroups);
+        // Check for duplicate IDs in the config
+        List<String> duplicateIdErrorMessages = this.vertices().parallel()
```

Comment:
@andrewvc As in the above, I don't think `parallel` get's us much here. 

Also, do we care about perfromance a lot here? If so, we should probably just move away from all the functional magic here, it's really slow in Java :) If we're happy with the performance now I'd just drop the `parallel()` to kill some dead code :)